lib/src/test/java/edu/stanford/futuredata/macrobase/analysis/classify/QuantileClassifierTest.java

```diff
@@ -0,0 +1,120 @@
+import org.junit.Test;
+
+import java.util.*;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class QuantileClassifierTest {
+    private int length = 100;
+    private DataFrame df;
+    private double[] quantiles;
+    private List<String> quantileColumnNames;
+    private LinkedHashMap<String, Double> quantileColumnsMap = new LinkedHashMap<String, Double>() {{
+        put("q0.0", 0.0);
+        put("q0.1", 0.1);
+        put("q0.5", 0.5);
+        put("q0.9", 0.9);
+        put("q1.0", 1.0);
+    }};
+    private double[] rawData;
+
+    @Before
+    public void setUp() {
+        df = new DataFrame();
+        double[] counts = new double[length];
+        double[] means = new double[length];
+        int c = 0;
+        quantileColumnNames = new ArrayList<String>();
+        quantiles = new double[quantileColumnsMap.size()];
+        for (Map.Entry<String, Double> entry : quantileColumnsMap.entrySet()) {
+            quantileColumnNames.add(entry.getKey());
+            quantiles[c++] = entry.getValue();
+        }
+
+        rawData = new double[length*101];
+        int d = 0;
+        List<double[]> quantileColumns = new ArrayList<double[]>();
+        for (int i = 0; i < quantiles.length; i++) {
+            quantileColumns.add(new double[length]);
+        }
+        for (int i = 0; i < length; i++) {
+            for (int j = i-50; j <= i+50; j++) {
+                rawData[d++] = j;
+            }
+            counts[i] = 101;
+            means[i] = i;
+            for (int j = 0; j < quantiles.length; j++) {
+                quantileColumns.get(j)[i] = i + (quantiles[j] - 0.5) * 100;
+            }
+        }
+        df.addDoubleColumn("count", counts);
+        df.addDoubleColumn("mean", means);
+        for (int i = 0; i < quantiles.length; i++) {
+            df.addDoubleColumn(quantileColumnNames.get(i), quantileColumns.get(i));
+        }
+    }
+
+    @Test
+    public void testClassify() throws Exception {
+        assertEquals(length, df.getNumRows());
+        QuantileClassifier ac = new QuantileClassifier("count", "mean", quantileColumnsMap);
+        ac.process(df);
+        DataFrame output = ac.getResults();
+        assertEquals(df.getNumRows(), output.getNumRows());
+        assertEquals(7, df.getSchema().getNumColumns());
+        assertEquals(8, output.getSchema().getNumColumns());
+
+        Percentile percentile = new Percentile();
+        percentile.setData(rawData);
+        double trueLowCutoff = percentile.evaluate(1);
+        double trueHighCutoff = percentile.evaluate(99);
+        assertEquals(trueLowCutoff, ac.getLowCutoff(), 5.0);
+        assertEquals(trueHighCutoff, ac.getHighCutoff(), 5.0);
+
+        double[] outliers = output.getDoubleColumnByName("_OUTLIER");
+
+        for (int i = 0; i < outliers.length; i++) {
+            int trueNumOutliers = 0;
+            trueNumOutliers += Math.max(0, trueLowCutoff - (i-50));
```

Comment:
This math gets very confusing: if we just save rawdata it's really easy to count how many are actually above / below the threshold. That also makes the test more robust to when you change the data generation logic.

Reply:
Yeah, I just wasn't sure if we wanted to have to save a List of arrays, in order to save all the raw data for each group. That would be in addition to the array of raw data that I'm already saving. Or I could just save the List of arrays and reconstruct the array of raw data by unrolling the List of arrays, but that seemed like a bunch of unnecessary logic and copying.